File: src/main/java/the_fireplace/netheressence/NetherEssence.java
package the_fireplace.netheressence;

import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.client.Minecraft;
import net.minecraft.client.resources.model.ModelResourceLocation;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.Loader;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.Mod.Instance;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.registry.GameRegistry;
import the_fireplace.netheressence.blocks.BlockNetherEssence;
import the_fireplace.netheressence.blocks.BlockRadiantNetherEssence;
import the_fireplace.netheressence.fulcrumcompat.FulcrumCompat;
import the_fireplace.netheressence.fulcrumcompat.IFulcrumCompat;
import the_fireplace.netheressence.handlers.NetherEssenceFuelHandler;

@Mod(modid = NetherEssence.MODID, name = NetherEssence.MODNAME, version = NetherEssence.VERSION)
public class NetherEssence
{
	@Instance(NetherEssence.MODID)
	public static NetherEssence instance;

	public static final String MODID = "netheressence";
	public static final String MODNAME = "Nether Essence";
	public static final String VERSION = "2.1.1.0";
	public static final String downloadURL = "http://goo.gl/zF0elV";

	public static final CreativeTabs tabNetherEssence = new CreativeTabs("tabNetherEssence"){
		@Override
		public Item getTabIconItem() {
			return NetherEssence.nether_essence;
		}
	};
	public static final Item nether_essence = new Item().setCreativeTab(NetherEssence.tabNetherEssence).setUnlocalizedName("NetherDust");
	public static final Block nether_essence_block = new BlockNetherEssence(Material.rock);
	public static final Block radiant_nether_essence_block = new BlockRadiantNetherEssence(Material.rock);

	@EventHandler
	public void preInit(FMLPreInitializationEvent event) {
		GameRegistry.registerBlock(nether_essence_block, "NetherDustBlock");
		GameRegistry.registerItem(nether_essence, "NetherDust");
		GameRegistry.registerBlock(radiant_nether_essence_block, "radiant_nether_essence_block");
		GameRegistry.registerFuelHandler(new NetherEssenceFuelHandler());
		IFulcrumCompat c;
		if(Loader.isModLoaded("fulcrum")){
			c=new FulcrumCompat();
			c.register();
		}
	}

	@EventHandler
	public void load(FMLInitializationEvent event) {
		ItemStack dustStack = new ItemStack(nether_essence);
		ItemStack waterStack = new ItemStack(Items.water_bucket);
		ItemStack soulStack = new ItemStack(Blocks.soul_sand);
		ItemStack netherrackStack = new ItemStack(Blocks.netherrack);
		ItemStack coalStack = new ItemStack(Items.coal);
		ItemStack redStack = new ItemStack(Items.redstone);
		ItemStack potionStack = new ItemStack(Items.potionitem);
		ItemStack glowDustStack = new ItemStack(Items.glowstone_dust);

		GameRegistry.addRecipe(new ItemStack(nether_essence, 4), "xxx", "yzy", "xxx",
				'x', netherrackStack, 'y', soulStack, 'z', waterStack);
		GameRegistry.addRecipe(new ItemStack(nether_essence, 4), "xyx", "xzx", "xyx",
				'x', netherrackStack, 'y', soulStack, 'z', potionStack);
		GameRegistry.addRecipe(new ItemStack(nether_essence, 4), "xyx", "xzx", "xyx",
				'x', netherrackStack, 'y', soulStack, 'z', waterStack);
		GameRegistry.addRecipe(new ItemStack(nether_essence, 4), "xxx", "yzy", "xxx",
				'x', netherrackStack, 'y', soulStack, 'z', potionStack);
		GameRegistry.addRecipe(new ItemStack(Items.gunpowder), "drd", "rcr", "drd",
				'd', dustStack, 'r', redStack, 'c', coalStack);
		GameRegistry.addRecipe(new ItemStack(Items.gunpowder), "rdr", "dcd", "rdr",
				'd', dustStack, 'r', redStack, 'c', coalStack);
		GameRegistry.addRecipe(new ItemStack(nether_essence_block), "xxx", "x x", "xxx",
				'x', dustStack);
		GameRegistry.addRecipe(new ItemStack(radiant_nether_essence_block), "xxx", "xgx", "xxx",
				'x', dustStack, 'g', glowDustStack);
		GameRegistry.addShapelessRecipe(new ItemStack(nether_essence, 8), new ItemStack(nether_essence_block));
		GameRegistry.addShapelessRecipe(new ItemStack(nether_essence, 8), new ItemStack(radiant_nether_essence_block));
		if(event.getSide().isClient()){
			registerItemRenders();
		}
	}
	public void registerItemRenders(){
		Minecraft.getMinecraft().getRenderItem().getItemModelMesher()
		.register(net.minecraft.item.Item.getItemFromBlock(nether_essence_block), 0,
				new ModelResourceLocation(NetherEssence.MODID+":NetherDustBlock", "inventory"));
		Minecraft.getMinecraft().getRenderItem().getItemModelMesher()
		.register(net.minecraft.item.Item.getItemFromBlock(radiant_nether_essence_block), 0,
				new ModelResourceLocation(NetherEssence.MODID+":NetherDustBlock", "inventory"));
		Minecraft.getMinecraft().getRenderItem().getItemModelMesher()
		.register(nether_essence, 0,
				new ModelResourceLocation(NetherEssence.MODID+":NetherDust", "inventory"));
	}
}

File: src/main/scala/the_fireplace/netheressence/NetherEssence.scala
package the_fireplace.netheressence

import java.lang.{Character => JChar}

import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.init.{Blocks, Items}
import net.minecraft.item.{Item, ItemBlock, ItemStack}
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.Mod.EventHandler
import net.minecraftforge.fml.common.event.{FMLInitializationEvent, FMLPreInitializationEvent}
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import net.minecraftforge.registries.IForgeRegistry
import the_fireplace.netheressence.blocks.{BlockItemRecovery, BlockNetherEssence, BlockRadiantNetherEssence}

@Mod(modid = NetherEssence.MODID, name = NetherEssence.MODNAME, modLanguage = "scala", acceptedMinecraftVersions = "[1.12,1.13)")
@Mod.EventBusSubscriber
object NetherEssence {
	final val MODID = "netheressence"
	final val MODNAME = "Nether Essence"

	val tabNetherEssence: CreativeTabs = new CreativeTabs("tabNetherEssence") {
		override def createIcon: ItemStack = new ItemStack(NetherEssence.nether_essence)
	}

	val nether_essence = new Item().setCreativeTab(NetherEssence.tabNetherEssence).setTranslationKey("NetherDust").setRegistryName("NetherDust")

	val nether_essence_block = new BlockNetherEssence(Material.ROCK).setRegistryName("NetherDustBlock")

	val radiant_nether_essence_block = new BlockRadiantNetherEssence(Material.ROCK).setRegistryName("radiant_nether_essence_block")

	val item_recovery_block = new BlockItemRecovery(Material.ROCK).setRegistryName("item_recovery_block")

	var blockRegistry: IForgeRegistry[Block] = _

	def registerBlock(block: Block) {
		if (blockRegistry == null)
			return
		blockRegistry.register(block)
	}

	var itemRegistry: IForgeRegistry[Item] = _

	def registerItem(item: Item) {
		if (itemRegistry == null)
			return
		itemRegistry.register(item)
	}

	def registerItemForBlock(block: Block) {
		if (itemRegistry == null)
			return
		itemRegistry.register(new ItemBlock(block).setRegistryName(block.getRegistryName))
	}

	@SubscribeEvent
	def registerBlocks(event: RegistryEvent.Register[Block]) {
		blockRegistry = event.getRegistry
		registerBlock(nether_essence_block)
		registerBlock(radiant_nether_essence_block)
		registerBlock(item_recovery_block)
	}

	@SubscribeEvent
	def registerItems(event: RegistryEvent.Register[Item]) {
		itemRegistry = event.getRegistry
		registerItem(nether_essence)
		registerItemForBlock(nether_essence_block)
		registerItemForBlock(radiant_nether_essence_block)
		registerItemForBlock(item_recovery_block)
	}

	@SideOnly(Side.CLIENT)
	@SubscribeEvent
	def registerItemRenders(event: ModelRegistryEvent) {
		ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(nether_essence_block), 0, new ModelResourceLocation(NetherEssence.MODID + ":NetherDustBlock", "inventory"))
		ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(item_recovery_block), 0, new ModelResourceLocation(NetherEssence.MODID + ":item_recovery_block", "inventory"))
		ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(radiant_nether_essence_block), 0, new ModelResourceLocation(NetherEssence.MODID + ":radiant_nether_essence_block", "inventory"))
		ModelLoader.setCustomModelResourceLocation(nether_essence, 0, new ModelResourceLocation(NetherEssence.MODID + ":NetherDust", "inventory"))
	}
}
